### PR TITLE
Let round_capa for ID table not allocate excess capacity for power of 2 ints >= 4

### PR DESCRIPTION
**Pull request by methodmissing**
While looking into allocation patterns of ID table I noticed cases of a capacity requested for a power of 2 value initializes an items array with the size rounded up to the next power of 2 value.

* `round_capa(4)` -> returns `8`
* `round_capa(8)` -> returns `16`
* `round_capa(16)` -> returns `32`
etc.

I'm not sure if this is a desired side effect of the ID table implementation, but it seems wasteful to allocate the extra items capacity for this case, given that the ruby heap objects with an ID table reference would be oldgen and thus the malloc heap overhead of over allocating ID table capacity would be pretty much permanent. This PR changes that to:

* `round_capa(4)` -> returns `4`
* `round_capa(8)` -> returns `8`
* `round_capa(16)` -> returns `16`
etc.

#### Memory savings

Not spectacular, but also not nothing. The probed values for `Class`, `Module` and `Object` are repeatable across various console sessions (up until Redmine booted, no work done yet):

* `Class` shaved `38208` bytes
* `Module` shaved `51520` bytes
* `Object` shaved `51572` bytes

`141300` bytes or circa `137` kilobytes for the above 3 types. Pending evaluation on a large Rails application that has a lot more method table churn than just up to an application booted stage.

Redmine post boot, trunk:

```
lourens@CarbonX1:~/src/redmine$ bundle exec rails c
Loading development environment (Rails 5.2.1.1)
irb(main):001:0> require 'objspace'
=> true
irb(main):002:0> GC.start
=> nil
irb(main):003:0> RUBY_DESCRIPTION
=> "ruby 2.7.0dev (2019-02-17 trunk 67081) [x86_64-linux]"
irb(main):004:0> ObjectSpace.each_object(Class).sum {|a| ObjectSpace.memsize_of(a) }
=> 2971888
irb(main):005:0> ObjectSpace.each_object(Module).sum {|a| ObjectSpace.memsize_of(a) }
=> 3851576
irb(main):006:0> ObjectSpace.each_object(Object).sum {|a| ObjectSpace.memsize_of(a) }
=> 22299833
```

Redmine post boot, this PR:

```
lourens@CarbonX1:~/src/redmine$ bundle exec rails c
Loading development environment (Rails 5.2.1.1)
irb(main):001:0> require 'objspace'
=> true
irb(main):002:0> GC.start
=> nil
irb(main):003:0> RUBY_DESCRIPTION
=> "ruby 2.7.0dev (2019-02-17 id-table-round.. 67081) [x86_64-linux]"
irb(main):004:0> ObjectSpace.each_object(Class).sum {|a| ObjectSpace.memsize_of(a) }
=> 2933680
irb(main):005:0> ObjectSpace.each_object(Module).sum {|a| ObjectSpace.memsize_of(a) }
=> 3800056
irb(main):006:0> ObjectSpace.each_object(Object).sum {|a| ObjectSpace.memsize_of(a) }
=> 22248261
```

Subset of core benchmarks with the memory runner comes in mostly at a little lighter as well (`compare-ruby` is trunk, `built-ruby` is this PR):

```
lourens@CarbonX1:~/src/ruby/ruby$ /usr/local/bin/ruby --disable=gems -rrubygems -I./benchmark/lib ./benchmark/benchmark-driver/exe/benchmark-driver             --executables="compare-ruby::~/src/ruby/trunk/ruby -I.ext/common --disable-gem"             --executables="built-ruby::./miniruby -I./lib -I. -I.ext/common  -r./prelude --disable-gem"   -r memory          $(ls ./benchmark/**.{yml,rb})
Calculating -------------------------------------
                                                              compare-ruby  built-ruby 
                                                   app_answer      11.432M     11.244M bytes -       1.000 times
                                                  app_aobench      11.908M     11.948M bytes -       1.000 times
                                                      app_erb      12.264M     11.816M bytes -     15.000k times
                                                app_factorial      23.496M     23.972M bytes -       1.000 times
                                                      app_fib      11.384M     11.272M bytes -       1.000 times
                                              app_lc_fizzbuzz      29.588M     24.792M bytes -       1.000 times
                                               app_mandelbrot      11.624M     11.492M bytes -       1.000 times
                                                app_pentomino      11.872M     11.708M bytes -       1.000 times
                                                    app_raise      11.484M     11.412M bytes -       1.000 times
                                                app_strconcat      11.292M     11.316M bytes -       1.000 times
                                                      app_tak      11.360M     11.184M bytes -       1.000 times
                                                    app_tarai      11.444M     11.280M bytes -       1.000 times
                                                      app_uri      12.508M     12.252M bytes -       1.000 times
                                       array_sample_100k__100      16.904M     17.236M bytes -       1.000 times
                                      array_sample_100k___10k      81.436M     81.344M bytes -       1.000 times
                                         array_sample_100k_10      13.500M     13.148M bytes -       1.000 times
                                         array_sample_100k_11      13.516M     13.384M bytes -       1.000 times
                                        array_sample_100k__1k      30.232M     31.288M bytes -       1.000 times
                                      array_sample_100k___50k      82.268M     82.160M bytes -       1.000 times
                                        array_sample_100k__6k      81.404M     81.132M bytes -       1.000 times
                                                  array_shift       1.574G      1.574G bytes -       1.000 times
                                              array_small_and      11.620M     11.564M bytes -       1.000 times
                                             array_small_diff      12.032M     11.612M bytes -       1.000 times
                                               array_small_or      11.876M     11.724M bytes -       1.000 times
                                             array_sort_block      28.444M     29.676M bytes -       1.000 times
                                             array_sort_float      28.540M     29.724M bytes -       1.000 times
                                          array_values_at_int      11.492M     11.324M bytes -       1.000 times
                                        array_values_at_range      80.824M     85.792M bytes -       1.000 times
                                                      bighash     306.176M    306.080M bytes -       1.000 times
                                            complex_float_add      11.456M     11.336M bytes -      1.000M times
                                            complex_float_div      11.392M     11.196M bytes -      1.000M times
                                            complex_float_mul      11.368M     11.244M bytes -      1.000M times
                                            complex_float_new      11.316M     11.248M bytes -      1.000M times
                                          complex_float_power      11.896M     11.740M bytes -      1.000M times
                                            complex_float_sub      11.368M     11.112M bytes -      1.000M times
                                                  dir_empty_p      12.308M     12.132M bytes -       1.000 times
                                         enum_lazy_grep_v_100      11.496M     11.256M bytes -       1.000 times
                                          enum_lazy_grep_v_20      11.488M     11.288M bytes -       1.000 times
                                          enum_lazy_grep_v_50      11.480M     11.332M bytes -       1.000 times
                                           enum_lazy_uniq_100      24.460M     23.780M bytes -       1.000 times
                                            enum_lazy_uniq_20      14.496M     13.768M bytes -       1.000 times
                                            enum_lazy_uniq_50      20.448M     19.792M bytes -       1.000 times
                                                   erb_render      13.452M     13.352M bytes -      1.500M times
                                                  fiber_chain      66.380M     64.112M bytes -       1.000 times
                                                   file_chmod      12.476M     12.052M bytes -       1.000 times
                                                  file_rename      12.508M     12.224M bytes -       1.000 times
                                          hash_aref_dsym_long     119.076M    119.892M bytes -       1.000 times
                                               hash_aref_dsym      11.360M     11.236M bytes -       1.000 times
                                                hash_aref_fix      11.340M     11.240M bytes -       1.000 times
                                                hash_aref_flo      12.052M     11.876M bytes -       1.000 times
                                               hash_aref_miss      11.380M     11.324M bytes -       1.000 times
                                                hash_aref_str      11.392M     11.228M bytes -       1.000 times
                                           hash_aref_sym_long      11.408M     11.132M bytes -       1.000 times
                                                hash_aref_sym      11.436M     11.172M bytes -       1.000 times
                                                 hash_flatten      67.408M     67.324M bytes -       1.000 times
                                               hash_ident_flo      11.948M     11.748M bytes -       1.000 times
                                               hash_ident_num      11.424M     11.224M bytes -       1.000 times
                                               hash_ident_obj      11.416M     11.304M bytes -       1.000 times
                                               hash_ident_str      11.364M     11.268M bytes -       1.000 times
                                               hash_ident_sym      11.436M     11.264M bytes -       1.000 times
                                                    hash_keys      85.880M     85.748M bytes -       1.000 times
                                          hash_literal_small2     300.724M    298.808M bytes -       1.000 times
                                          hash_literal_small4     300.504M    298.784M bytes -       1.000 times
                                          hash_literal_small8     300.552M    298.788M bytes -       1.000 times
                                                    hash_long      11.388M     11.224M bytes -       1.000 times
                                                   hash_shift      12.508M     12.348M bytes -       1.000 times
                                               hash_shift_u16      15.796M     15.584M bytes -       1.000 times
                                               hash_shift_u24      15.780M     15.656M bytes -       1.000 times
                                               hash_shift_u32      15.864M     15.588M bytes -       1.000 times
                                                  hash_small2     298.788M    298.616M bytes -       1.000 times
                                                  hash_small4     300.828M    298.668M bytes -       1.000 times
                                                  hash_small8     300.820M    298.760M bytes -       1.000 times
                                                 hash_to_proc      11.912M     11.712M bytes -       1.000 times
                                                  hash_values      85.936M     85.848M bytes -       1.000 times
                                                      int_quo      11.392M     11.208M bytes -       1.000 times
                                         io_copy_stream_write      11.428M     11.320M bytes -       1.000 times
                                  io_copy_stream_write_socket      12.028M     11.504M bytes -       1.000 times
                                               io_file_create      11.768M     11.740M bytes -       1.000 times
                                                 io_file_read      38.620M     37.604M bytes -       1.000 times
                                                io_file_write      12.388M     12.324M bytes -       1.000 times
                                            io_nonblock_noex2      11.308M     11.336M bytes -       1.000 times
                                             io_nonblock_noex      12.564M     12.620M bytes -       1.000 times
                                                   io_pipe_rw      11.484M     11.252M bytes -       1.000 times
                                                   io_select2      11.724M     11.512M bytes -       1.000 times
                                                   io_select3      13.356M     13.004M bytes -       1.000 times
                                                    io_select      11.260M     11.084M bytes -       1.000 times
                                                     loop_for      11.444M     11.224M bytes -       1.000 times
                                               loop_generator      11.384M     11.288M bytes -       1.000 times
                                                   loop_times      11.404M     11.204M bytes -       1.000 times
                                              loop_whileloop2      11.356M     11.104M bytes -       1.000 times
                                               loop_whileloop      11.476M     11.208M bytes -       1.000 times
                                             marshal_dump_flo      18.216M     18.060M bytes -       1.000 times
                                      marshal_dump_load_geniv      11.536M     11.484M bytes -       1.000 times
                                       marshal_dump_load_time      11.608M     11.656M bytes -       1.000 times
                                     (1..1_000_000).last(100)      11.412M     11.112M bytes -       1.000 times
                                    (1..1_000_000).last(1000)      11.380M     11.264M bytes -       1.000 times
                                   (1..1_000_000).last(10000)      11.336M     11.384M bytes -       1.000 times
                                               require_thread      12.592M     12.212M bytes -       1.000 times
                                                      require      28.168M     27.772M bytes -       1.000 times
                                                 securerandom      11.448M     11.444M bytes -       1.000 times
                                                 so_ackermann      11.680M     11.660M bytes -       1.000 times
                                                     so_array      11.544M     11.456M bytes -       1.000 times
                                              so_binary_trees      37.148M     28.924M bytes -       1.000 times
                                               so_concatenate     128.608M    128.408M bytes -       1.000 times
                                               so_count_words      18.788M     19.152M bytes -       1.000 times
                                                 so_exception      11.812M     11.660M bytes -       1.000 times
                                                  so_fannkuch      11.948M     11.840M bytes -       1.000 times
                                                     so_fasta      13.340M     13.548M bytes -       1.000 times
                                              so_k_nucleotide      42.212M     42.036M bytes -       1.000 times
                                                     so_lists      64.140M     63.692M bytes -       1.000 times
```

#### Other changes

* Moved the `POW2_P` macro from Bignum out to `internal.h`
* Injected that into `vm_method.c` and removed the duplicate `POWER_OF_2_P` macro

Thoughts much appreciated if there's a better way to go about with behavior `round up to current or next power of 2, minimum 4` capacity as opposed to the current `round up to next power of 2, minimum 4` (assuming that's desired off course and I'm not missing context).

#### Not yet investigated

* Impact on a very large Rails application
* I see no noticeable performance regression and `round_capa` is not in a very hot path, but can do more testing around this if required.

### CODE REVIEW

File: id_table.c [hunk 1]
```diff
@@ -69,6 +69,8 @@ ITEM_SET_KEY(struct rb_id_table *tbl, int i, id_key_t key)
 static inline int
 round_capa(int capa)
 {
+    /* let power of 2 ints >= 4 not increase capa */
+    if ((capa >= 4) && POW2_P(capa)) return capa;
```
Comment (ahorek) [inline id_table.c]: would you mind to test https://github.com/ruby/ruby/pull/2086/files as an alternative?

**Comment by ahorek** [comment]
`round_capa(4)` -> returns `8`

maybe @ko1 knows, but it looks like an unintentional side effect to me.

**Comment by methodmissing** [comment]
Thanks for investigating @ahorek - tried several other approaches the weekend too and most had either the same behaviour or worse side effects. I think @nobu may have some insights too.

**Comment by methodmissing** [comment]
References discussion in https://bugs.ruby-lang.org/issues/15631

**Comment by methodmissing** [comment]
@k0kubun any objections to reopening this PR again?

**Comment by k0kubun** [comment]
No. I'm even not sure why I'm asked here, because what I did were only refreshing the base branch and merging master to have `check_branch` job to unblock merge.

**Comment by methodmissing** [comment]
Sorry for the ping - didn't realize the last commits were refreshes only.

**Pull request closed without merging**